src/Core/Resolvers/SqlMutationEngine.cs

```diff
@@ -369,38 +404,43 @@ public class SqlMutationEngine : IMutationEngine
+                // Even if its first result, postgresql may still be an insert op here, if so, return CreatedResult
+                if (_sqlMetadataProvider.GetDatabaseType() is DatabaseType.PostgreSQL)
                 {
-                    Dictionary<string, object?> resultRow = dbResultSetRow.Columns;
-
-                    bool isUpdateResultSet = false;
-                    if (upsertOperationResult.ResultProperties.TryGetValue(IS_UPDATE_RESULT_SET, out object? isUpdateResultSetValue))
-                    {
-                        isUpdateResultSet = Convert.ToBoolean(isUpdateResultSetValue);
-                    }
+                    isUpdateResultSet = !PostgresQueryBuilder.IsInsert(resultRow);
+                }
 
-                    // For MsSql, MySql, if it's not the first result, the upsert resulted in an INSERT operation.
-                    // Even if its first result, postgresql may still be an insert op here, if so, return CreatedResult
-                    if (!isUpdateResultSet ||
-                        (_sqlMetadataProvider.GetDatabaseType() is DatabaseType.PostgreSQL &&
-                        PostgresQueryBuilder.IsInsert(resultRow)))
+                // When read permissions is configured without database policy, a subsequent select query will not be executed.
+                // However, the read action could have include and exclude fields configured and it
+                // Any additional fields that are present in the response are removed.
+                if (isReadPermissionConfiguredForRole && !isDatabasePolicyDefinedForReadAction)
+                {
+                    HashSet<string> allowedExposedColumns = _authorizationResolver.GetAllowedExposedColumns(context.EntityName, roleName, EntityActionOperation.Read).ToHashSet();
+                    foreach (string columnInResponse in resultRow.Keys)
                     {
-                        string primaryKeyRoute = ConstructPrimaryKeyRoute(context, resultRow);
-                        // location will be updated in rest controller where httpcontext is available
-                        return new CreatedResult(location: primaryKeyRoute, OkMutationResponse(resultRow).Value);
+                        if (!allowedExposedColumns.Contains(columnInResponse))
+                        {
+                            resultRow.Remove(columnInResponse);
+                        }
                     }
+                }
 
-                    // Valid REST updates return OkObjectResult
-                    return OkMutationResponse(resultRow);
+                // When the upsert operation results in the creation of a new record, a HTTP 201 CreatedResult response is returned.
+                if (!isUpdateResultSet)
+                {
+                    return ConstructCreatedResultResponse(resultRow, selectOperationResponse, string.Empty, isReadPermissionConfiguredForRole, isDatabasePolicyDefinedForReadAction);
```

Comment:
A comment here as to why the PK is left as empty for location header would help.

Reply:
Added